src/stan/mcmc/chains.hpp

```diff
@@ -254,22 +256,39 @@ namespace stan {
           var_plus += variance(chain_mean);
         Eigen::VectorXd rho_hat_t(n_samples);
         rho_hat_t.setZero();
-        double rho_hat = 0;
-        int max_t = 0;
-        for (int t = 1; (t < n_samples && rho_hat >= 0); t++) {
-          Eigen::VectorXd acov_t(chains);
-          for (int chain = 0; chain < chains; chain++) {
-            acov_t(chain) = acov(chain)(t);
+        double rho_hat_even = 1;
+        double rho_hat_odd = 0;
+        Eigen::VectorXd acov_t(chains);
+        for (int chain = 0; chain < chains; chain++)
+          acov_t(chain) = acov(chain)(1);
+        rho_hat_odd = 1 - (mean_var - mean(acov_t)) / var_plus;
+        rho_hat_t(1) = rho_hat_odd;
+        // Geyer's initial positive sequence
+        int max_t = 1;
+        for (int t = 1;
+             (t < (n_samples - 2) && (rho_hat_even + rho_hat_odd) >= 0);
+             t += 2) {
+          for (int chain = 0; chain < chains; chain++)
+            acov_t(chain) = acov(chain)(t+1);
+          rho_hat_even = 1 - (mean_var - mean(acov_t)) / var_plus;
+          for (int chain = 0; chain < chains; chain++)
+            acov_t(chain) = acov(chain)(t+2);
+          rho_hat_odd = 1 - (mean_var - mean(acov_t)) / var_plus;
+          if ((rho_hat_even + rho_hat_odd) >= 0) {
+            rho_hat_t(t+1) = rho_hat_even;
+            rho_hat_t(t+2) = rho_hat_odd;
           }
-          rho_hat = 1 - (mean_var - mean(acov_t)) / var_plus;
-          if (rho_hat >= 0)
-            rho_hat_t(t) = rho_hat;
-          max_t = t;
+          max_t = t+2;
         }
-        double ess = chains * n_samples;
-        if (max_t > 1) {
-          ess /= 1 + 2 * rho_hat_t.sum();
+        // Geyer's initial monotone sequence
+        for (int t = 3; t <= (max_t - 2); t += 2) {
+          if ((rho_hat_t(t+1) + rho_hat_t(t+2)) > (rho_hat_t(t-1) + rho_hat_t(t))) {
+            rho_hat_t(t+1) = (rho_hat_t(t-1) + rho_hat_t(t)) / 2;
+            rho_hat_t(t+2) = rho_hat_t(t+1);
+          }
```

Comment:
I believe there's an error here. If you're only using the initial monotone sequence, then I believe you need (not sure how to describe this better):
```
else {
  rho_hat_t(t + 1) = 0;
  rho_hat_t(t + 2) = 0;
}
```
all the way out to `max_t`. I believe if the monotone sequence length is less than `max_t`, the elements beyond the monotone sequence still has possibly non-zero values in it.

The alternative is to wrap this condition into the loop above this one and only go out as long as it is monotone.

Reply:
monotone sequence length is `max_t`

Reply:
thanks @syclik , see two of my comments inline. I'll update tests